### ISSUE TITLE
Cache cannot be shared in concurrent queries

### ISSUE THREAD

**Issue by kyri-petrou** [zio/zio-query#450]
Simple repro:

```scala
import zio.*
import zio.query.*

object Main extends ZIOAppDefault {
  case object Req extends Request[Nothing, Unit]

  val ds: DataSource[Any, Req.type] =
    DataSource.fromFunctionZIO("MyDatasource")((_: Req.type) => ZIO.unit.delay(50.millis))

  def run =
    for {
      cache <- Cache.empty
      _     <- ZIO.foreachParDiscard(1 to 100)(_ => ZQuery.fromRequest(Req)(ds).runCache(cache))
    } yield ()
}
```

Throws:

```scala
Exception in thread "zio-fiber-5" zio.query.QueryFailure: Data source MyDatasource did not complete request Req.
```

I understand that this is not how the cache is meant to be shared between concurrent queries, but there are some cases where we can't run the queries in parallel via `ZQuery.foreachPar`. One use-case where we need to run queries in parallel while sharing the cache is when we process queries in a `ZStream`. In Caliban, we stream GraphQL `@defer` fields back to the client. The queries in the stream need to share the same cache otherwise we lose the ability to deduplicate requests to the same DataSource.

As a workaround, we are forced to process each effect in the `ZStream` sequentially (see https://github.com/ghostdogpr/caliban/pull/1981 and attached issue). However, this doesn't perform well when we have multiple deferred responses.

### Potential solution(?)

One potential solution to the issue would be for the Cache to use `Promise`s instead of `Ref`s. I did an experiment (see https://github.com/zio/zio-query/commit/034e5e58bfb469b6688c782258e1e372321dee20), and the approach seems to work. 

However, I'm not sure whether there might be edge-cases where we'd be causing deadlocks, as well as what the performance implication of using Promises over Refs is. Any thoughts? Is this something that could potentially be of use?

**Comment by adamgfraser** [comment]
This is definitely not the intended use of the cache. We can explore supporting this, but to some extent I think this is the wrong approach as it gives you caching but not batching.

I think a better approach would be to change the `run` method in your `Executor` to not return a `ZStream` but instead preserve the original structure of a query that can potentially return another query. Then you can use `ZQuery.foreachPar` to run the first part of all the queries, batching them appropriately, and then if those queries generate further queries run those in the same way with the updated cache, constructing a stream out of that.

**Comment by kyri-petrou** [comment]
@adamgfraser This case is quite particular and we're happy to forgo batching. Fields annotated by the client as `@defer` are not meant to be included in the initial response, but instead streamed back in a multipart response when each finishes processing. They are meant to be used for data that is expensive to calculate but not needed for the initial page load (see [here](https://www.apollographql.com/docs/react/data/defer/) for more info).

Therefore we can't / shouldn't batch execute them as some of them might require a lot of time to process, but we still want to be able to cache them.

**Comment by kyri-petrou** [comment]
Thanks @adamgfraser 🥇
